graph/src/components/subgraph/proof_of_indexing/online.rs

```diff
@@ -0,0 +1,207 @@
+}
+
+impl fmt::Debug for ProofOfIndexing {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("ProofOfIndexing").field(&"...").finish()
+    }
+}
+
+impl ProofOfIndexing {
+    pub fn new(block_number: u64) -> Self {
+        Self {
+            block_number,
+            per_causality_region: HashMap::new(),
+        }
+    }
+    /// Adds an event to the digest of the ProofOfIndexingStream local to the DataSource
+    pub fn write(
+        &mut self,
+        logger: &Logger,
+        causality_region: &str,
+        event: &ProofOfIndexingEvent<'_>,
+    ) {
+        if *LOG_EVENTS {
+            debug!(
+                logger,
+                "Proof of indexing event";
+                "event" => &event,
+                "causality_region" => causality_region
+            );
+        }
+
+        // This may be better with the raw_entry API, once that is stabilized
+        if let Some(causality_region) = self.per_causality_region.get_mut(causality_region) {
+            causality_region.write(event);
+        } else {
+            let mut entry = BlockEventStream::new(self.block_number);
+            entry.write(event);
+            self.per_causality_region
+                .insert(causality_region.to_owned(), entry);
+        }
+    }
+    pub fn take(self) -> HashMap<String, BlockEventStream> {
+        self.per_causality_region
+    }
+}
+
+pub struct ProofOfIndexingFinisher {
+    block_number: u64,
+    state: SetHasher,
+    causality_count: usize,
+}
+
+impl ProofOfIndexingFinisher {
+    pub fn new(
+        block: &EthereumBlockPointer,
+        subgraph_id: &SubgraphDeploymentId,
+        indexer_address: &Option<H256>,
+    ) -> Self {
+        let mut state = SetHasher::new();
+
+        // Add the subgraph id
+        let subgraph_id_seq_no = traverse_seq_no(&[
+            1, // PoI.subgraph_id
+        ]);
+        subgraph_id.stable_hash(subgraph_id_seq_no, &mut state);
+
+        // Add the block number
+        let block_no_seq_no = traverse_seq_no(&[
+            2, // PoI.block_number
+        ]);
+        block.stable_hash(block_no_seq_no, &mut state);
+
+        // Add the indexer
+        let indexer_seq_no = traverse_seq_no(&[
+            3, // PoI.indexer_address
+        ]);
+        let indexer_bytes = indexer_address.as_ref().map(|i| AsBytes(i.as_bytes()));
+        indexer_bytes.stable_hash(indexer_seq_no, &mut state);
+
+        ProofOfIndexingFinisher {
+            block_number: block.number,
+            state,
+            causality_count: 0,
+        }
+    }
+
+    pub fn add_causality_region(&mut self, name: &str, region: &[u8]) {
+        let mut state = SetHasher::from_bytes(region);
+
+        // Finish the blocks vec
+        let blocks_seq_no = traverse_seq_no(&[
+            1, // kvp -> v
+            0, // CausalityRegion.blocks: Vec<Block>
+        ]);
+        // + 1 is for
+        (self.block_number + 1).stable_hash(blocks_seq_no, &mut state);
```

Comment:
Great question. That is redundant. I had the `block_ptr` field be of type EthereumBlockPointer - so that's where one of the block numbers came from. The other (this one, with the `+ 1` comment) is to delineate the end of the `events` Vec on `Block<'a>`. I'll fix it to just change the `block_ptr` field to `block_hash` and fix the comment.